data_preparation/AgML_predictor_data_prepare.py

```diff
@@ -0,0 +1,1077 @@
+import os, sys,time
+import numpy as np
+import rasterio
+import rasterio.features
+import pandas as pd
+import geopandas as gpd
+from rasterio.crs import CRS
+from itertools import repeat
+from multiprocessing import Pool
+import logging
+from zipfile import ZipFile
+from rasterio.warp import transform_geom
+from shapely.geometry import shape
+from rasterio.mask import mask
+from datetime import datetime
+import multiprocessing
+import datetime
+import warnings
+warnings.filterwarnings("ignore")
+
+
+log = logging.getLogger(__name__)
+
+
+SUPPRESS_ERRORS = True
+TIFF_DIRECTORY_NAME = "Tiff"
+EU_COUNTRY_CODE_KEY = "CNTR_CODE"
+EU_ADMIN_LEVEL_KEY = "LEVL_CODE"
+
+# Country codes to admin level
+# Austria (AT), Belgium (BE), Bulgaria (BG), Czech Republic (CZ), Germany (DE), Denmark (DK), Estonia (EE), Greece (EL),
+# Spain (ES), Finland (FI), France (FR), Croatia (HR), Hungary (HU), Ireland (IE), Italy (IT), Lithuania (LT),
+# Latvia (LV), The Netherlands (NL), Poland (PL), Portugal (PT), Romania (RO), Sweden (SE), Slovakia (SK)
+EU_COUNTRIES = {
+    "AT" : 2,
+    "BE" : 2,
+    "BG" : 2,
+    "CZ" : 3,
+    "DE" : 3,
+    "DK" : 3,
+    "EE" : 3,
+    "EL" : 3,
+    "ES" : 3,
+    "FI" : 3,
+    "FR" : 3,
+    "HR" : 2,
+    "HU" : 3,
+    "IE" : 2,
+    "IT" : 3,
+    "LT" : 3,
+    "LV" : 3,
+    "NL" : 2,
+    "PL" : 2,
+    "PT" : 2,
+    "RO" : 3,
+    "SE" : 3,
+    "SK" : 3
+}
+
+# Angola (AO), Burkina Faso (BF), Ethiopia (ET), Lesotho (LS), Madagascar (MG), Malawi (MW),
+# Mozambique (MZ), Niger (NE), Senegal (SN), Chad (TD), South Africa (ZA), Zambia (ZM)
+FEWSNET_COUNTRIES = ["AO", "BF", "ET", "LS", "MG", "MW", "MZ", "NE", "SN", "TD", "ZA", "ZM"]
+FEWSNET_ADMIN_ID_KEY = "adm_id"
+
+
+"""
+@author: Joint Research Center
+"""
+
+class UnableToExtractStats(Exception):
+    pass
+
+"""
+@author: Joint Research Center
+"""
+def read_masked(ds, mask, indexes=None, window=None,
+                use_pixels='CENTER', out_mask_path=None,
+                *args, **kwargs):
+    
+    """
+    Reads the data fom the raster file and returns it as a numpy masked array.
+    The geometry is used to define the mask.
+    It returns only a subset of the data.
+        It reads only the raster window containing the geometry. This optimizes the performance by reducing disk reads.
+    The "use_pixels" parameter can be used to define how to handle the border pixels.
+        Some require shapely to be available on the system
+    :param ds: Raster file path or an instance of a rasterio.DatasetReader (already opened raster file).
+    :param mask: geojson geometry used to mask the array, mask is a read mask (only data covered by the geometry is used)
```

Comment:
The description should be updated to match the code change from the original. "mask" changed form geometry -> iterable over geometries

Reply:
Hi Pvojnov and Dilli, I do not have time to test the new version (which iterable over geometries). Please use this version.

Reply:
@umdsgy : The idea is to review, address and test. If you cannot test, then the code will stay as a pull request and never merged to main.

Reply:
@umdsgy, this comment was just to update the function description to match the changes you made to the original code.
@krsnapaudel, I believe that  the code in this PR is probably tested and that @umdsgy is referring to my general comment to try to use the additional part of the code that was shared after there was some complaints that the execution is slow.

Reply:
yes, @pvojnov is on the same page with me. The current code has been tested and run for months.

Reply:
@umdsgy: If it's tested, please share some results from the run. They need to match the format of data in the benchmark dataset.